tools/classes/AssetsManager.ts

```diff
@@ -0,0 +1,438 @@
+
+		return result;
+	}
+
+	async getCdnAssets(): Promise<{
+		logo: string | false;
+		thumbnail: string | false;
+		assets: Map<number, string> | false;
+	}> {
+		const assets = new Map<number, string>();
+		let assetFound = true,
+			index = 0;
+		while (assetFound) {
+			const asset = await this.doesAssetExistAnyExtension(
+				`${this.assetBaseUrl}/${index}`
+			);
+			if (asset) {
+				assets.set(index, asset);
+				index++;
+			} else {
+				assetFound = false;
+			}
+		}
+
+		return {
+			logo: await this.doesAssetExistAnyExtension(`${this.assetBaseUrl}/logo`),
+			thumbnail: await this.doesAssetExistAnyExtension(
+				`${this.assetBaseUrl}/thumbnail`
+			),
+			assets: assets.size ? assets : false,
+		};
+	}
+
+	canBePut(oldUrl: string, newUrl: string): boolean {
+		return this.getFileExtension(oldUrl) === this.getFileExtension(newUrl);
+	}
+
+	async doesAssetExist(url: string): Promise<boolean> {
+		return got
+			.head(url)
+			.then(() => true)
+			.catch(() => false);
+	}
+
+	async doesAssetExistAnyExtension(url: string): Promise<string | false> {
+		const extensions = [".png", ".jpg", ".jpeg", ".gif", ".webp"];
+		for (const extension of extensions) {
+			const newUrl = `${url}${extension}`;
+			if (await this.doesAssetExist(newUrl)) return newUrl;
+		}
+		return false;
+	}
+
+	async uploadAssets(assets: Map<string, string>) {
+		let errors: string[] = [];
+		for (const [url, newUrl] of assets) {
+			const extension = this.getFileExtension(url);
+			let mimeType: string;
+			switch (extension) {
+				case ".png":
+					mimeType = "image/png";
+					break;
+				case ".jpg":
+				case ".jpeg":
+					mimeType = "image/jpeg";
+					break;
+				case ".gif":
+					mimeType = "image/gif";
+					break;
+				case ".webp":
+					mimeType = "image/webp";
+					break;
+				default:
+					errors.push(
+						`Tried to upload an asset with an invalid extension: ${url}`
+					);
+					continue;
+			}
+
+			const random = Math.random().toString(36).substring(2, 15),
+				filename = `premid-assetmanager-${random}${extension}`,
+				fileLocation = join(tmpdir(), filename);
+
+			let finalFileLocation = fileLocation;
+
+			try {
+				const stream = got.stream(url);
+				await pipeline(stream, createWriteStream(fileLocation));
+			} catch (error) {
+				errors.push(`Error while downloading asset ${url}: ${error.message}`);
+				continue;
+			}
+
+			if (!newUrl.includes("thumbnail")) {
+				const file = sharp(fileLocation),
+					metadata = await file.metadata();
+
+				if (metadata.width !== 512 || metadata.height !== 512) {
+					try {
+						const newFileLocation = join(tmpdir(), `resized-${filename}`);
+						await file
+							.resize(512, 512, {
+								fit: "contain",
+								background: { r: 0, g: 0, b: 0, alpha: 0 },
+							})
+							.toFile(newFileLocation);
+						finalFileLocation = newFileLocation;
+					} catch (error) {
+						errors.push(`Error while resizing asset ${url}: ${error.message}`);
+						continue;
+					}
+				}
+			}
```

Comment:
Maybe we should do the same for thumbnails (With a different sizing proportion tho? Idk how we would prevent stretching tho)